src/types/acls.rs

```diff
@@ -0,0 +1,351 @@
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
+}
+
+impl Acl {
+    /// Parse the given input into a [`ACL`] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        AclTryBuilder {
+            data: owned,
+            acl_builder: |input| {
+                let mut lines: &[u8] = input;
+
+                // There should only be ONE single ACL response
+                while !lines.is_empty() {
+                    match imap_proto::parser::parse_response(lines) {
+                        Ok((_rest, Response::Acl(a))) => {
+                            // lines = rest;
+                            return Ok(InnerAcl {
+                                mailbox: a.mailbox,
+                                acls: a
+                                    .acls
+                                    .into_iter()
+                                    .map(|e| AclEntry {
+                                        identifier: e.identifier,
+                                        rights: e.rights.into(),
+                                    })
+                                    .collect(),
+                            });
+                        }
+                        Ok((rest, data)) => {
+                            lines = rest;
+                            if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                                return Err(resp.into());
+                            }
+                        }
+                        _ => {
+                            return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+                        }
+                    }
+                }
+
+                Err(Error::Parse(ParseError::Invalid(lines.to_vec())))
+            },
+        }
+        .try_build()
+    }
+
+    pub fn mailbox(&self) -> &str {
+        &*self.borrow_acl().mailbox
+    }
+
+    pub fn acls(&self) -> &[AclEntry<'_>] {
+        &*self.borrow_acl().acls
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct InnerAcl<'a> {
+    pub(crate) mailbox: Cow<'a, str>,
+    /// The list of identifier/rights pairs for the mailbox
+    pub(crate) acls: Vec<AclEntry<'a>>,
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The list of identifiers and rights for the ACL response
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclEntry<'a> {
+    /// The user identifier the rights are for
+    pub identifier: Cow<'a, str>,
+    /// the rights for the provided identifier
+    pub rights: AclRightList,
+}
+
+/// From [section 3.7 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.7).
+///
+/// The LISTRIGHTS response from the listrights IMAP command
+#[self_referencing]
+pub struct ListRights {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) rights: InnerListRights<'this>,
```

Comment:
I don't think you need this inner type here either — you can just declare the fields directly here.